kconfig: Rename CONFIG_FLOAT to CONFIG_FPU

stephanosio opened the pull request:
```
This commit renames the Kconfig `FLOAT` symbol to `FPU`, since this
symbol only indicates that the hardware Floating Point Unit (FPU) is
used and does not imply and/or indicate the general availability of
toolchain-level floating point support (i.e. this symbol is not
selected when building for an FPU-less platform that supports floating
point operations through the toolchain-provided software floating point
library).

Moreover, given that the symbol that indicates the availability of FPU
is named `CPU_HAS_FPU`, it only makes sense to use "FPU" in the name of
the symbol that enables the FPU.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

There was a related discussion at https://github.com/zephyrproject-rtos/zephyr/pull/23948#issuecomment-606430554.

### Q&A/Rationale

**Q:** Doesn't `CONFIG_FLOAT` mean that floating point support in general (i.e. not just FPU) is enabled?
**A:** Just looking at its name, it is very reasonable for one to think that; but, that is not the case. This symbol is used by: 1) arch implementations to enable and configure the hardware FPU, 2) tests to determine whether to run hardware FPU-dependent tests; therefore, it should be called `CONFIG_FPU`.

**Q:** In that case, what about the symbol that specifies whether floating point support, in general, is enabled?
**A:** To clarify, there was no such symbol to begin with; the semantic of `CONFIG_FLOAT` has always been `CONFIG_FPU`, and any usage of `CONFIG_FLOAT` to infer general floating point support would be incorrect.

**Q:** Shouldn't we add a separate symbol to specify whether floating point support in general (FPU or software) is _available_?
**A:** Non-hardware FPU floating point support is always software-based, and its availability is dependent on the toolchain being used. For instance, in case of GCC, when an FPU-less target is specified, software floating point support is automatically enabled and there is no further user intervention necessary to get floating point support; if software floating point support is not available for the specified target (notably, AArch64), [all `float` and `double` usages will produce compiler errors](https://gist.github.com/stephanosio/02ad346af5d02e9a93f0228b853039f2). Noting this, we could potentially define a toolchain symbol to specify whether software floating point support is available (e.g. `TOOLCHAIN_HAS_SOFT_FP`) in the future, but there does not seem to be an immediate need to do so.

**Q:** Shouldn't we add a separate symbol to specify whether floating point usage is _enabled_?
**A:** This is really an application/library-level specification and should be left for the application to decide. As a general rule of thumb, kernel and "core OS" code should not make use of floating point features unless hardware FPU is enabled.

stephanosio committed:
kconfig: Rename CONFIG_FLOAT to CONFIG_FPU

This commit renames the Kconfig `FLOAT` symbol to `FPU`, since this
symbol only indicates that the hardware Floating Point Unit (FPU) is
used and does not imply and/or indicate the general availability of
toolchain-level floating point support (i.e. this symbol is not
selected when building for an FPU-less platform that supports floating
point operations through the toolchain-provided software floating point
library).

Moreover, given that the symbol that indicates the availability of FPU
is named `CPU_HAS_FPU`, it only makes sense to use "FPU" in the name of
the symbol that enables the FPU.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

carlescufi reviewed (approved):
Full marks for the PR description, complete with an extensive FAQ. I say you take parts of that text and add it to the `menuconfig` help section.

pfalcon commented:
Does this PR somehow address concerns of https://github.com/zephyrproject-rtos/zephyr/issues/14122 (which was closed unresolved)?

stephanosio commented:
> Does this PR somehow address concerns of #14122 (which was closed unresolved)?

@pfalcon No, I did not touch the symbol description since this PR focuses on renaming only and the current description does align with the semantic of `CONFIG_FPU`. I will see if I can address that here.

> Full marks for the PR description, complete with an extensive FAQ. I say you take parts of that text and add it to the `menuconfig` help section.

@carlescufi I will add an additional commit that reworks the symbol description for `CONFIG_FPU` and `CONFIG_FP_SHARING`.

pabigot reviewed (approved):
I'd swear when I wrote https://github.com/zephyrproject-rtos/zephyr/pull/23948#issuecomment-606565206 that I was seeing code that used `CONFIG_FLOAT` to determine whether to calculate a sensor value with integer or FP operands, but I can't find it now, so this seems good.

stephanosio committed:
kconfig: Improve architecture floating point symbol descriptions

This commit reworks the symbol descriptions for `CONFIG_FPU` and
`CONFIG_FP_SHARING`, in order to provide more details and clarify any
ambiguity between the two symbols.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Added a commit to improve the descriptions for `CONFIG_FPU` and `CONFIG_FP_SHARING`.

![image](https://user-images.githubusercontent.com/11345418/80177575-1319c280-8637-11ea-8d97-cf11f55e219a.png)

![image](https://user-images.githubusercontent.com/11345418/80177644-3e9cad00-8637-11ea-9fd7-64f8fde28fda.png)

pfalcon commented on tests/benchmarks/app_kernel/src/master.h:
```diff
@@ -23,7 +23,7 @@
 
 /* uncomment the define below to use floating point arithmetic */
-/* #define FLOAT */
+/* #define FPU */
```
I don't think that this (and related changes) are correct. This is local define (`FLOAT`) of  tests/benchmarks/app_kernel, not directly related to Zephyr config option `CONFIG_FLOAT`. And it's meaning seems to be exactly use floating-point arithmetic (which can be also soft-float), rather than specifically FPU.

So, I'd suggest these changes to be reverted.

stephanosio replied:
Good catch, I will revert these.

stephanosio replied:
Reverted

pfalcon reviewed (changes requested):
@stephanosio, I appreciate taking care of improving FPU-related config option description in Kconfig.

On the other hand, I think changes applied to tests/benchmarks/app_kernel/ are superfluous and ultimately confusing. Given that this PR collected enough +1's, let me give -1 to avoid premature merging, before this issue is clarified. Thanks.

pfalcon reviewed (approved):
Thanks!

pfalcon commented:
@stephanosio 

> Added a commit to improve the descriptions for CONFIG_FPU and CONFIG_FP_SHARING

Once again, I appreciate taking care of that. Now if we look at those 2 options - `CONFIG_FPU` and `CONFIG_FP_SHARING` - and read their description, it would come clear that "CONFIG_FP_SHARING" should actually be "CONFIG_FP**U**_SHARING".

And I would really hate to add more scope creep to your work, especially given that I couldn't find suitable window to do that for a year (re: #14122). But the common problem is exactly finding enough time to perform a comprehensive enough analysis. And you seem to be right in the context of it. On the plus side, if you agree that addition of one letter can largely improve consistency and clarity (which IMHO it does), then we'd limit all the breaking changes to config options to the next Zephyr release (2.3), and would have ~ perfect picture in that area afterwards and unlikely need to disturb users again.

And actually, if'd you agree with that, I may very well suggest to make the change as a separate PR, after this one is merged. Surely, no problem if you don't consider it to be worth it.

Thanks!

stephanosio commented:
> it would come clear that "CONFIG_FP_SHARING" should actually be "CONFIG_FP**U**_SHARING".

@pfalcon While I agree that it would be more clear and straight forward to prefix the symbol with `_FPU_` instead of `_FP_`, if we are to take this approach, we would need to change many others as well to be consistent:

_Definitely:_
* `CONFIG_EAGER_FP_SHARING` -> `CONFIG_EAGER_FPU_SHARING`
* `CONFIG_LAZY_FP_SHARING` -> `CONFIG_LAZY_FPU_SHARING`

_Possibly:_
* `K_FP_REGS` -> `K_FPU_REGS`
* `_FP_USER_MASK` -> `_FPU_USER_MASK`
* `DO_FP_CALLER_SAVED` -> `DO_FPU_CALLER_SAVED`
* `DO_FP_CALLEE_SAVED` -> `DO_FPU_CALLEE_SAVED`
* `STORE_FP_CALLER_SAVED` -> `STORE_FPU_CALLER_SAVED`
* `STORE_FP_CALLEE_SAVED` -> `STORE_FP_CALLEE_SAVED`

I believe many would find these changes "rather subjective" and "not essential."

**I will probably consider opening a separate PR for these changes if this comment gets a reasonable amount of thumbs up.**

pfalcon commented:
>     CONFIG_EAGER_FP_SHARING -> CONFIG_EAGER_FPU_SHARING
>    CONFIG_LAZY_FP_SHARING -> CONFIG_LAZY_FPU_SHARING

I didn't even know such options exist. Ideally, they would have been called CONFIG_FPU_SHARING_EAGER, etc., to make it clear they're sub-options of CONFIG_FPU_SHARING. Oh no, I don't call to put the adjective at the end now. But I agree they should have "FPU" in them.

> K_FP_REGS -> K_FPU_REGS
> ...

Well, the talk was about external interface of Zephyr configuration system, config options name, and them being self-describing (including relationship among options). I wouldn't extend this proposed refactor to (apparently internal) symbol names used in the source code. It's definitely a different scope.


Thanks for considering it, voted up!
